lib/slack/client.ex

```diff
@@ -0,0 +1,257 @@
+defmodule Slack.Client do
+  @moduledoc """
+  A datastructure that allows you to monitor, inspect, and react to the state of
+  your Slack team.
+  """
+  
+  if :lt == Version.compare(System.version, "1.1.0") do
+    
+    @derive [Access]
+    
+  else
+  
+    @behaviour Access
+    
+    @type key :: any
+    @type value :: any
+
+    @doc """
+    Implements fetch/2 for Access protocol.
+    """
+    @spec fetch(Client.t, key) :: value
+    def fetch(client, key)
+    defdelegate fetch(client, key), to: Map
+      
+    @doc """
+    Implements get_and_update/3 for Access protocol.
+    """
+    @spec get_and_update(Client.t, key, (value -> {value, value})) :: Client.t
+    def get_and_update(client, key, function)
+    defdelegate get_and_update(client, key, function), to: Map
+    
+  end
+  
+  defstruct [
+    socket: nil,
+    client: nil,
+    token:  nil,
+    me:     nil,
+    team:   nil,
+    bots:     %{},
+    channels: %{},
+    groups:   %{},
+    users:    %{},
+    ims:      %{},
+  ]
+    
+  @doc """
+  Notice when a new Slack object has been created.
+  
+  Available types: `[:channel, :im, :group, :user, :bot]`
+  """
+  @spec track(Client.t, Atom.t, Map.t) :: Client.t
+  def track(client, type, object)
+  
+  [:channel, :im, :group, :user, :bot] |> Enum.map( fn type ->
+    plural_type = type |> Atom.to_string |> Kernel.<>("s") |> String.to_atom
+    def track(client, unquote(type), object) do
+      put_in client, [unquote(plural_type), object.id], object
+    end
+  end )
+  
+  @doc """
+  Notices when you join a channel or group.
+  
+  Available types: `[:channel, :group]`
+  """
+  @spec join(Client.t, Atom.t, Map.t) :: Client.t
+  def join(client, type, object)
+  
+  def join(client, :channel, channel) do
+    client
+      |> put_in([:channels, channel.id, :members], channel.members)
+      |> put_in([:channels, channel.id, :is_member], true)
+  end
+  
+  def join(client, :group, group) do
+    put_in(client, [:groups, group.id], group)
+  end
+  
+  @doc """
+  Notices when a user joins a channel or group.
+  
+  Available types: `[:channel, :group]`
+  """
+  @spec joined(Client.t, Atom.t, Map.t, Map.t) :: Client.t
+  def joined(client, type, object, user)
+  
+  [:channel, :group] |> Enum.map( fn type ->
```

Comment:
I would do the same thing here that I suggested above for `track`. We could probably do without the module attributes after seeing this.